### PR TITLE
planner: revise isnullRejected check for `And` and `OR`

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/38304

Problem Summary:

The `And` and `OR` result would be changed by Null which is transformed by the column schema in `isNullRejected`

### What is changed and how it works?

revise isnullRejected check for `And` and `OR`

### Check List

Tests 

- [x] Unit test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: expression/constant_fold.go [hunk 1]
```diff
@@ -181,6 +181,11 @@ func foldConstant(expr Expression) (Expression, bool) {
 			if !hasNullArg || !sc.InNullRejectCheck || x.FuncName.L == ast.NullEQ {
 				return expr, isDeferredConst
 			}
+			// For `AND` and `OR`, if any of the arguments is null and not all arguments are constant, then it is unFoldAble.
```
Comment (xuyifangreeneyes) [inline expression/constant_fold.go:184]:
```suggestion
			// For `AND` and `OR`, if any of the arguments is null and not all arguments are constant, then it is unFoldable.
```

Comment (xuyifangreeneyes) [reply expression/constant_fold.go:184]: Maybe we can add examples to explain why.

**Review by xuyifangreeneyes — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**